dart/lib/flat_buffers.dart

```diff
@@ -1241,3 +1246,72 @@ class _VTable {
     }
   }
 }
+
+// Allocator interface. This is flatbuffers-specific - only for [Builder].
+abstract class Allocator {
+  ByteData allocate(int size);
+
+  void deallocate(ByteData data);
+
+  /// Reallocate [newSize] bytes of memory, replacing the old [oldData]. This
+  /// grows downwards, and is intended specifically for use with [Builder].
+  /// Params [inUseBack] and [inUseFront] indicate how much of [oldData] is
+  /// actually in use at each end, and needs to be copied.
+  ByteData reallocateDownward(
+      ByteData oldData, int newSize, int inUseBack, int inUseFront) {
+    final newData = allocate(newSize);
+    clear(newData, true);
+    _copyDownward(oldData, newData, inUseBack, inUseFront);
+    deallocate(oldData);
+    return newData;
+  }
+
+  /// Clear the allocated data contents.
+  ///
+  /// Param [isFresh] is true if the given data has been freshly allocated,
+  /// depending on the allocator implementation, clearing may be unnecessary.
+  void clear(ByteData data, bool isFresh) {
+    final length = data.lengthInBytes;
+    final length64b = (length / 8).floor();
+    // fillRange iterates over data so it's faster with larger data type
+    if (length64b > 0) data.buffer.asUint64List().fillRange(0, length64b, 0);
+    data.buffer.asUint8List().fillRange(length64b * 8, length % 8, 0);
+  }
+
+  /// Called by [reallocate] to copy memory from [oldData] to [newData]. Only
+  /// memory of size [inUseFront] and [inUseBack] will be copied from the front
+  /// and back of the old memory allocation.
+  void _copyDownward(
+      ByteData oldData, ByteData newData, int inUseBack, int inUseFront) {
+    if (inUseBack != 0) {
+      newData.buffer.asUint8List().setAll(
+          newData.lengthInBytes - inUseBack,
+          oldData.buffer.asUint8List().getRange(
+              oldData.lengthInBytes - inUseBack, oldData.lengthInBytes));
+    }
+    if (inUseFront != 0) {
+      newData.buffer
+          .asUint8List()
+          .setAll(0, oldData.buffer.asUint8List().getRange(0, inUseFront));
+    }
+  }
+}
+
+class DefaultAllocator extends Allocator {
+  @override
+  ByteData allocate(int size) => ByteData(size);
+
+  @override
+  void deallocate(ByteData _) {
+    // nothing to do, it's garbage-collected
+  }
+
+  @override
+  void clear(ByteData data, bool isFresh) {
+    if (isFresh) {
+      // nothing to do, ByteData is created all zeroed out
+    } else {
+      super.clear(data, isFresh);
```

Comment:
It seems strange that we'd let a holder of the buffer or the allocator clear things.

Reply:
I agree, but it may be the most performant way (it is, significantly so, with the given C-backed allocator)

Reply:
What I mean is - you can clear the buffer either by using `allocator.clear` or `buffer.clear` or some other method directly on the buffer. I don't think it has any performance implications either way, and if we just have a custom buffer class we can enforce where clear is defined and document what we expect it to do.

Reply:
IOW, I'm wondering if we really want `Allocator` to just be a factory for creating new `BuilderBuffer` instances or something (and maybe resizing them?), and that `BuilderBuffer` (or whatever we'd name it) being responsible for having `dispose`, `clear`, etc.

Reply:
So, about the `BuilderBuffer` class. Wouldn't that mean that whoever wants to use a custom allocation actually has to implement the whole `ByteData` interface, even if it's just for forwarding to another object already implementing `ByteData`? Or what would be the signature of the `BuilderBuffer` class you had in mind?

Reply:
We can just make it have the methods we care aboutand create a trivial wrapper around ByteBuffer for it

Reply:
Something like this?

```dart

/// The interface that [Builder] uses to allocate buffers for encoding.
abstract class Allocator {
  const Allocator();

  /// Allocate a buffer of a given size.
  BuilderBuffer allocate(int size);

  BuilderBuffer resize(
      BuilderBuffer oldData, int newSize, int inUseBack, int inUseFront) {
    final newData = allocate(newSize);
    newData.clear();
    _copyDownward(oldData, newData, inUseBack, inUseFront);
    oldData.dispose();
    return newData;
  }

  void _copyDownward(BuilderBuffer oldData, BuilderBuffer newData,
      int inUseBack, int inUseFront) { /* see committed code */  }
}

/// Buffer used for [Builder], created by a [Allocator] implementation.
abstract class BuilderBuffer {
  /// Get access to the underlying data
  ByteData get data;

  /// Free the buffer after it's not going to be used by the [Builder] anymore.
  ///
  /// For example: deallocate any memory used for the buffer.
  void dispose();

  /// Clear the contents - zeroing out all bytes.
  void clear();
}
```

Reply:
If so, maybe resize could also move to the BuilderBuffer.

Reply:
in which case the allocator could actually be just a single function, not a class